pkg/controller/proxyconfig/validation.go

```diff
@@ -0,0 +1,346 @@
+
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/validation"
+)
+
+const (
+	// proxyProbeMaxRetries is the number of times to attempt an http GET
+	// to a readinessEndpoints endpoint.
+	proxyProbeMaxRetries = 3
+	// clusterConfigMapKey is the name of the data key containing the
+	// PEM-encoded CA trust bundle.
+	clusterConfigMapKey = "ca-bundle.crt"
+	proxyHTTPScheme     = "http"
+	proxyHTTPSScheme    = "https"
+	// noProxyWildcard is the string used to as a wildcard attached to a
+	// domain suffix in proxy.spec.noProxy to bypass proxy for httpProxy
+	// and httpsProxy.
+	noProxyWildcard = "*"
+	// certPEMBlock is the type taken from the preamble of a PEM-encoded structure.
+	certPEMBlock = "CERTIFICATE"
+)
+
+// ValidateProxyConfig ensures that proxyConfig is valid.
+func (r *ReconcileProxyConfig) ValidateProxyConfig(proxyConfig *configv1.ProxySpec) error {
+	if specHTTPProxySet(proxyConfig) {
+		scheme, err := validateURI(proxyConfig.HTTPProxy)
+		if err != nil {
+			return fmt.Errorf("invalid httpProxy URI: %v", err)
+		}
+		if scheme != proxyHTTPScheme {
+			return fmt.Errorf("httpProxy requires an %q URI scheme", proxyHTTPScheme)
+		}
+	}
+
+	var readinessCerts []*x509.Certificate
+	if specHTTPSProxySet(proxyConfig) {
+		scheme, err := validateURI(proxyConfig.HTTPSProxy)
+		if err != nil {
+			return fmt.Errorf("invalid httpsProxy URI: %v", err)
+		}
+		if scheme == proxyHTTPSScheme {
+			// A trusted CA bundle must be provided when using https
+			// between client and proxy.
+			if !specTrustedCASet(proxyConfig) {
+				return errors.New("trustedCA is required when using an https scheme with httpsProxy")
+			}
+			certBundle, err := r.validateTrustedCA(proxyConfig)
+			if err != nil {
+				return fmt.Errorf("failed validating TrustedCA %q: %v", proxyConfig.TrustedCA.Name, err)
+			}
+			copy(certBundle, readinessCerts[0:])
+		}
+	}
+
+	if specNoProxySet(proxyConfig) {
+		if proxyConfig.NoProxy != noProxyWildcard {
+			for _, v := range strings.Split(proxyConfig.NoProxy, ",") {
+				v = strings.TrimSpace(v)
+				errDomain := validateDomainName(v, false)
+				_, _, errCIDR := net.ParseCIDR(v)
+				if errDomain != nil && errCIDR != nil {
+					return fmt.Errorf("invalid noProxy: %v", v)
+				}
+			}
+		}
+	}
+
+	if specReadinessEndpoints(proxyConfig) {
+		for _, endpoint := range proxyConfig.ReadinessEndpoints {
+			scheme, err := validateURI(endpoint)
+			if err != nil {
+				return fmt.Errorf("invalid URI for endpoint %s: %v", endpoint, err)
+			}
+			switch {
+			case scheme == proxyHTTPScheme:
+				if err := validateHTTPReadinessEndpoint(endpoint); err != nil {
+					return fmt.Errorf("readinessEndpoint probe failed for endpoint %s", endpoint)
+				}
+			case scheme == proxyHTTPSScheme:
+				if !specTrustedCASet(proxyConfig) {
+					return fmt.Errorf("readinessEndpoint with an %q scheme requires trustedCA to be set", proxyHTTPSScheme)
```

Comment:
i don't think using an https proxy necessitates providing a trustedCA.  it's possible (perhaps unlikely) that your proxy is using a cert that's signed by a system-trust chain.

Reply:
Good point. I'll move the trusted ca check into the `isSpecTrustedCASet` block.

Reply:
@bparees I'm thinking through the implementation for your above comment. What would be the reason for providing a trustedCA to a proxy if not used for proxying https connections?

Reply:
well what i was stating was the opposite (one might use https proxy and not provide a CA).

but as for why one might provide a CA but not use https..   for proxy, no reason i can think of, but since we're discussing extending this to be a generic "provide CAs to components" feature, it's easy to envision that someone might have CAs and no https proxy, in that case.

Reply:
> one might use https proxy and not provide a CA

@bparees got it. I added the support for this requirement here: https://github.com/openshift/cluster-network-operator/pull/271/files#diff-1502a5b3db6bd911880e1d0d9346b561R47-R59